### ISSUE TITLE
chunk/split large files?

### ISSUE THREAD

**Issue by sjpotter** [rclone/rclone#497]
so I figure this might be out of scope for rclone, but figured I'd ask anyways.  Now that rclone is working for me with acd I can't maintain network connections for really large files.

might it be possible to add a file chunking to rclone (so that I can mark one of the repos as chunked so large files are written in small file chunks to a chunked destination, but converted to a whole if copied to a non chunked destination.

this way even for sites that don't support resemble uploads, one can have effective support for it.

<edited for a little clarity>

**Comment by ncw** [comment]
I see what you mean - nice idea.

**Comment by sjpotter** [comment]
Id also note that files under a certain size woulsnt have

**Comment by ncw** [comment]
This could be implemented using an overlay file system in a similar way to crypt

**Comment by ywLBhGk2noZLbY7** [comment]
My question is though, would the chunks be resumable? The current reason why I split my chunks is to resume my upload when _stuff_ hits the fan. If the chunks were resumable, and a customizable size (I.E. 1GB, which is what I currently do), then it'd be amazing as it'd save me the massive disk I/O of splitting them prior to uploading.

**Comment by ncw** [comment]
Yes the chunks would be resumable. The way it would work is as an overlay or modification of the local file system. Big files would appear as bigfile.1 bigfile.2 etc and these would sync separately. So if one part failed you could retry it.

**Comment by ermint** [comment]
It'd be more secure if it wasn't obvious which chunks made up a file.

It'd be even more secure if a local upload cache was maintained, with chunks uploaded in a random order, so even checking the order in which chunks are uploaded wouldn't reveal which blobs make up a single file.

It'd be great if the solution addressed #666 too.

**Comment by sjpotter** [comment]
ah, saw the comment about an overlay fs.  I agree it could be done that way, though another dependency and requires the same fs on all OSs one wants to use rclone with

**Comment by ShapeShifter499** [comment]
I'm looking forward to this because of the 50G Amazon Cloud file limit that seems to be in place.

**Comment by lolvince** [comment]
Hello, I come to the news regarding this function ...

**Comment by yatesco** [comment]
I'd _really_ like this but only if the block size was configurable (10MB blocks would work better with my poor bandwidth) and transparent, so, for example, it did nothing more than split at 10MB blocks, no obfuscating or mangling please :-). I think libraries that leave data-at-rest in an easily recoverable format are much more preferable to those where re-assembly requires the lib.

**Comment by sebitnt** [comment]
Hello everybody :-)
I'd just like to know if there has been some progress on this topic. Maybe there is a workaround available?  I want to upload mostly larger files to ACD and at the moment the transfers are failing at a ratio about 9/10. This is very unsatisfying for me because my internet connection is not very fast and it takes a long time to re upload the whole file.
Also I'd like to say THANK YOU for this great tool which did a great job for me apart from the described problems!

**Comment by mmiat** [comment]
+1

**Comment by Firemad03** [comment]
Hi,

It's a feature that rclone really need because of the limitation of the cloud.

I got a workaround that I will share with all of you. The key is the split command to create part file and the cat command to join them together when we want to decompress it in Linux/Unix. Split/cat command will do what we need but it will be better if rclone can support it natively.

Here my tough about my step:

1) Use tar to do a backup of my directories with the permissions
2) Use the compression LZ4 with tar to help compress the data the best we can
3) Secure it with GPG or OpenSSL , I know that rclone can crypt it. Why not make it secure first and use the crypt feature or rclone too ;)
4) Split will separate the crypted file to 1G for each file. You can use what you want. Just read the man for split if you want to tweak the command
5) Use cat to join the file together to uncrypt, decompress and untar it.

### Here my way to do it

```
### Encrypt - I called my file zfs2 because I like ZFS right ;)
tar cvf - zfs2 | lz4 - zfs2.tar.lz4 
gpg --output encrypted.data --symmetric --cipher-algo AES256 un_encrypted.data
split -a7 -d -b 1G - zfs2.tar.lz4.enc zfs2.tar.lz4.enc.part

Use rclone with crypt with a cloud storage :) Upload all the part.

### Decrypt
Use rclone to download the data
cat zfs2.tar.lz4.enc.part* >zfs2.tar.lz4.enc
gpg --output un_encrypted.data --decrypt encrypted.data
lz4 -d zfs2.tar.lz4 zfs2.tar
tar xvf zfs2.tar -C /mnt/RED01-Z3/Media/test
```

### Note to replace gpg with opensll
If for some reasons you can't use gpg because it's more secure then openssl that have a weakness here is the step with openssl:

```
### Encrypt
tar cvf - zfs2 | lz4 - zfs2.tar.lz4 
openssl aes-256-cbc -a -salt -in zfs2.tar.lz4 -out zfs2.tar.lz4.enc
split -a7 -d -b 1G - zfs2.tar.lz4.enc zfs2.tar.lz4.enc.part

Use rclone with crypt with a cloud storage :) Upload all the part.

### Decrypt
Use rclone to download the data
cat zfs2.tar.lz4.enc.part* >zfs2.tar.lz4.enc
openssl aes-256-cbc -d -a -in zfs2.tar.lz4.enc -out zfs2.tar.lz4
lz4 -d zfs2.tar.lz4 zfs2.tar
tar xvf zfs2.tar -C /mnt/RED01-Z3/Media/test
```

I'm sure maybe someone have a better way of doing it but it's working fine with rclone :) 

I tried with Freenas, FreeBSD of course, Centos, Cygwin (I suggest the babun shell and I'm using it in Windows just google it you will understand)

@sebitnt that's my workaround feel free to use what you need :)
@yatesco  you can tweak split with the 10M switch, change my 1G for 10M or anything you like

Please note that the openssl and gpg isn't required if you want to only use crypt. You can skip this step 👍

**Comment by Firemad03** [comment]
I really don't know if it's a good idea or not that rclone support it because I see rclone as an additional tool.

If you need to keep the permission in the cloud you need to tar it.
If you need to split file you need the split/cat command
If you need rsync for something it's rsync
If you need rclone for the cloud well it's rclone job.
If you need to encrypt the data you need an additional tool and I think rclone already crossed the line with crypt.

That's my opinion :) 

Off course I will like that rclone do everything easily but is it the goal of rclone project ?

**Comment by gordan-bobic** [comment]
Personally, I would find it very useful if chunk size was configurable down to 4-8KB in size, if it also allows for random read-write through the merged file (with read-modify amplification of up to 8KB per write). If the file on top could be made sparse (0-filled blocks not even created), that would be even better.

So if we create a chunked file foo of size 1GB with chunk size of 8KB, that would create:
foo/meta: (8192, 128)
foo/{001...128}: (none create pre-emptively, files created when that particular block is write-touched.

This would allow for all sorts of other awesome features, such as using rclone in a similar way to s3backer, where you can have a large file you can put a regular file system into. You could then have an advanced file system like ZFS on top of block storage in a very efficient way, and advantages of both.

**Comment by puzl** [comment]
I'd love to have this feature too.  the 5GB file size limit on box.com is difficult to work around.

**Comment by vivook** [comment]
Agree this should make impossible to statisicly analyse file size of encrypted files. If I upload movie from torrent site encrypted which has 42.827.372.823 bytes is easy to know what file because not lots of files that same size.

Could this put large files in chunks and upload in shuffled order with random names so cloud provider don't know which file chunks is which file? So above file is 42*1Gb + 1*827.372.823 uploaded in random order with 1Gb parts of other files. Is much more secure.

**Comment by TowerBR** [comment]
Another thing to evaluate: when I have to upload the big file again (say, a DB after some small modification), I'll have to upload all the small parts/chunks, or the algorithm will generate/select only the changed parts to be uploaded?

Upload just the changed parts would be a fantastic feature...

**Comment by Doomineer** [comment]
Look no further people, check out this project https://www.cryfs.org
Here's a comparison with other crypto-related projects https://nuetzlich.net/gocryptfs/comparison/
Hope this helps :)

**Comment by alpytron** [comment]
I would love this feature. Please implement it.

**Comment by jwr** [comment]
I would like to offer a real-life use case, to counter those comments that dismiss the problem.

I am trying to rclone a backup image, produced by SuperDuper. It's a ~400MB encrypted disk image (DMG), residing on an external USB-connected drive. I have a several-hour window at work when I can upload the image.

I tried yesterday: rclone spends 1.5 hours on an initial scan of the file (not even trying to upload anything). The upload to B2 did not complete, and it seems that rclone is unable to resume from an interrupted upload to B2.

Today I figured I'd split the DMG manually into 1GB chunks. I had to make space for another copy of the DMG. I started the split and that job has been running for the past 3 hours and it is about halfway done. And I haven't even got to the uploading part yet.

This is with a USB3-connected 3.5" drive, so it isn't the slowest thing possible.

I have a feeling that rclone (excellent as it is) could help a bit with this kind of use case.

**Comment by mlbarrow** [comment]
@jwr Because of the problem scenario you described, I whipped up a script to handle the use case of uploading a huge file. It is not perfect by any means, but allows you to get a big file uploaded without having to consume 2x the disk space.

https://github.com/mlbarrow/rclonechunker

```
$ ./rclonechunker
Missing filename
Usage: ./rclonechunker {filename} {rclone remote:path} [{chunk size in bytes}]
   chunk size defaults to 104857600

$ ./rclonechunker test.dat backblaze:mybucket/backup-files
Splitting test.dat into 5 chunks of 104857600 bytes and uploading to backblaze:mybucket/backup-files
Chunk #0...uploading
Chunk #1...uploading
Chunk #2...uploading
Chunk #3...uploading
Chunk #4...uploading
DONE!

$ ./rclonechunker test.dat backblaze:mybucket/backup-files
Splitting test.dat into 5 chunks of 104857600 bytes and uploading to backblaze:mybucket/backup-files
Chunk #0...skipping -- already present
Chunk #1...skipping -- already present
Chunk #2...skipping -- already present
Chunk #3...skipping -- already present
Chunk #4...skipping -- already present
DONE!
```

**Comment by jwr** [comment]
@mlbarrow Thanks, that's a useful script! It's also a perfect prototype for what I was hoping rclone would do (well, I also hoped it would re-assemble the file when getting it back).

**Comment by mlbarrow** [comment]
I am working on the downloader. Turns out that I have a similar issue as you do, but wasn’t motivated to deal with it. Thanks for the motivation!

**Comment by matt812** [comment]
Also, if you are able to split/rejoin files like this, many other things become possible.

For example, there is a current size limit of ~ 1MB per Google Docs document.
You get (basically) unlimited Google Docs documents for free (don't count towards Google Drive quota).
There is (or was?) a CLI-API for reading/writing those documents, too.

So, this seems to add up to a potentially unlimited amount of chunked, secured storage of any file types you want, so long as they are ASCII-armored inside of Google Docs:
- up: encrypt | ASCII-armor | split | gdoc-writes
- down: gdoc-reads | join | de-ASCII-armor | decrypt

Of course, all of this could be done with scripting (and without rclone, like with gpg and several unix tools/libraries), but having it packaged inside of an existing tool that already does encryption and API-interfacing would be really convenient.  And, that's why we use rclone -- convenience.  Rclone is basically a well-packaged set of script replacements (and gpg, cURL, etc.).

Just sayin'.

**Comment by DevPGSV** [comment]
I have a 200GB file in my account that I wanted to download with a slow connection. I know it's going to take days, but I can not keep my computer on for that long.

A possible solution (testing it right now, it seems alright) is the following:

`rclone serve http --addr 127.0.0.1:8888 remote:path/to/file.zip `
(Wait some seconds until rclone is set up)
`wget -v -c 127.0.0.1:8888/path/to/file.zip`

As wget with option "-c" continues the download from where it left, and appends the new data to the same (not yet complete) file, this can be executed several times in different sessions until the file is completely downloaded.

I'd like to suggest adding an option similar to wget "-c" for "rclone copy" and/or "rclone sync", so that when executed again they would continue the download insetad of starting it again.

**Comment by ncw** [comment]
@DevPGSV wrote:
> `rclone serve http --addr 127.0.0.1:8888 remote:path/to/file.zip`
> (Wait some seconds until rclone is set up)
> `wget -v -c 127.0.0.1:8888/path/to/file.zip`

Clever work-around :-)

> I'd like to suggest adding an option similar to wget "-c" for "rclone copy" and/or "rclone sync", so that when executed again they would continue the download insetad of starting it again.

A good idea... I wonder if I can work that into #2108

**Comment by dia3olik** [comment]
Hey Nick, the same principle applies to the option I suggested to resume transfers of big files after having interrupted it after a certain limit (to stay within google daily limits for example). For sure useful in downloads but maybe not in uploads since I'm not sure Google Drive and other hosters support resume on uploads...

**Comment by Riebart** [comment]
Not that we need _another_ script for chunking the contents of a pipe, but I wanted one that would do it in parallel to saturate higher bandwidth links, and let you do something arbitrary with the chunks (one of those things is rclone uploads): https://github.com/Riebart/pipechunker

**Comment by zenjabba** [comment]
Any update on getting around box.com 5GB limitations?

**Comment by zhendosina** [comment]
great idea, any news?

**Comment by cjreyn** [comment]
+1 for this

**Comment by ivandeex** [comment]
Hi Nick. I am preparing a PR, which implements the _chunker_ backend for rclone. I am still writing the documentation and working on the last failing integration test. Planning to submit the PR this month. In the meantime everybody interested can download beta binaries from the releases page. https://github.com/ivandeex/rclone

**Comment by ncw** [comment]
> Hi Nick. I am preparing a PR, which implements the _chunker_ backend for rclone. I am still writing the documentation and working on the last failing integration test. Planning to submit the PR this month. In the meantime everybody interested can download beta binaries from the releases page. https://github.com/ivandeex/rclone

That looks like a great bit of work - thank you :-)

Let me know if you need any help with the last bits and pieces.

I had some idea that we might auto wrap the destination if the `--chunk` flag was supplied - what do you think of that idea?  Or you could use `--chunker-remote remote: :chunker:`...

**Comment by ivandeex** [comment]
TL;DR I would be reluctant to add a new CLI flag.
First, having all chunking options packed in a single `--chunk` flag would put a rigid contract on the chunking wrapper (and more chunkers may appear in the future). I would first let the first chunker out in the wild and let use cases calm down.
Second, perhaps a `--chunker-remote` flag is OK as a shortcut, but we then logically come to a `--crypt-remote` and `--whatever-remote`. Why not stick to the single `--wrapper-remote` then? Also, if a user runs rclone with --chunker-remote=chunker1, and then with --chunker-remote=chunker2 with a diferent set of options, she would suddenly find the filesystem thrashed with previously hidden `whatever.chunker1.001` files. I would rather have this hardcoded in a more static manner in config file.
Last, rclone already has an overwhelming number of CLI flags all mixed in a single "namespace" from different backends and usage scenarios.

**Comment by ncw** [comment]
> TL;DR I would be reluctant to add a new CLI flag.
> First, having all chunking options packed in a single `--chunk` flag would put a rigid contract on the chunking wrapper (and more chunkers may appear in the future). I would first let the first chunker out in the wild and let use cases calm down.

Note that all backend options are available as flags, so the ones you've defined already will become flags automatically.

> Second, perhaps a `--chunker-remote` flag is OK as a shortcut, but we then logically come to a `--crypt-remote` and `--whatever-remote`.

A crypt setup from command line parameters is possible now.  It needs too many parameters to be sensible though.

> Why not stick to the single `--wrapper-remote` then? Also, if a user runs rclone with --chunker-remote=chunker1, and then with --chunker-remote=chunker2 with a diferent set of options, she would suddenly find the filesystem thrashed with previously hidden `whatever.chunker1.001` files. I would rather have this hardcoded in a more static manner in config file.

Fair enough. 

> Last, rclone already has an overwhelming number of CLI flags all mixed in a single "namespace" from different backends and usage scenarios.

It certainly isn't pretty!

There are 3 classes of flags, the ones for the command you are using, the global flags and the backend config - see https://rclone.org/flags/

**Comment by gordan-bobic** [comment]
How would this play with sparse files at source? Would this facilitate being able to upload a sparse file by not transferring chunks that are purely holes?

**Comment by ivandeex** [comment]
Fixed by https://github.com/rclone/rclone/pull/3503 and https://github.com/rclone/rclone/pull/3571

**Comment by ncw** [comment]
Thank you for implementing this @ivandeex :-)

I'm going to mark this ticket as done...

> Chunk resuming is not implemented, this needs a separate ticket.

Would you like to make another ticket about that?

**Comment by lbakyl** [comment]
Hi, I did not know that someone else was working on the the chunker option and it is great that it is integrated to rClone. So I created my own Python script that does it. Perhaps it could be useful for someone. I tested it on Synology, Mac OS X and Windows 10.
[https://github.com/lbakyl/rClone_chunking](url)
